### PR TITLE
Add template to test ipv6 and dual stack with k8s CI versions

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**

**What this PR does / why we need it**: Adds templates for ipv6 and dual stack using release-branch CI k8s versions so we can test various k8s release branches with CAPZ before releases to catch future regressions like https://github.com/kubernetes/kubernetes/issues/120999.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
@nojnhuh the broken link error is coming from https://github.com/nojnhuh/cluster-api-provider-azure/tree/aso/azure/services/asogroups which is referred to in the ASO proposal. I'm guessing that branch no longer exists... Is there a good replacement for it?

**Comment by nojnhuh** [comment]
The link now magically works again. I'll think about what better link to keep in the doc.
sry

**Comment by CecileRobertMichon** [comment]
This is ready for review

/assign @nawazkh @jackfrancis

**Comment by CecileRobertMichon** [comment]
/hold

will merge #3971 first and rebase

File: templates/test/ci/prow-ci-version-ipv6/patches/machine-deployment.yaml
Comment (nawazkh) [inline templates/test/ci/prow-ci-version-ipv6/patches/machine-deployment.yaml:38]: Minor nit: EOF new line.

**Review by nawazkh — APPROVED** [review]
Just a minor nit of EOF which is not a blocker tho, can be fixed in another PR to avoid running all the tests again.
Looks good to me! 🚢

File: templates/test/ci/prow-ci-version-dual-stack/patches/machine-deployment.yaml
Comment (willie-yao) [inline templates/test/ci/prow-ci-version-dual-stack/patches/machine-deployment.yaml:23]: nit: extra newline

File: test/e2e/conformance_test.go
Comment (willie-yao) [inline test/e2e/conformance_test.go:124]: Quick clarifying question: Will this still find the right flavor if we use CI or PR artifacts? I don't see pre-existing templates for them even without the ip suffix so I'm wondering if this may cause problems if we're getting the templates from elsewhere.

The flavors being set: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/0151252433f9b0d284312fe0dc761173c4980d4a/test/e2e/conformance_test.go#L115

File: templates/test/ci/prow-ci-version-ipv6/patches/machine-deployment.yaml
Comment (CecileRobertMichon) [reply templates/test/ci/prow-ci-version-ipv6/patches/machine-deployment.yaml]: fixed

File: templates/test/ci/prow-ci-version-dual-stack/patches/machine-deployment.yaml
Comment (CecileRobertMichon) [reply templates/test/ci/prow-ci-version-dual-stack/patches/machine-deployment.yaml]: fixed

File: test/e2e/conformance_test.go
Comment (CecileRobertMichon) [reply test/e2e/conformance_test.go:124]: Yes for CI, that's what this PR is enabling, the flavors are defined here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4086/files#diff-35f348a5d567ebeba39b1e40bbab0e1a7f64009cbadff142952f9b9691e94666R121

For PR artifacts no since those templates haven't been added. I could change the line so that it only changes the suffix for CI artifacts but then if you set "ipv6" and "pr artifacts" the test would silently fail (it would just use the regular template) instead of failing with a flavor not found error so I think it's arguably worse. This also is more future proof if we do add the equivalent ipv6/dual stack PR artifacts templates later.

**Comment by CecileRobertMichon** [comment]
/hold

rebase is done

File: test/e2e/data/kubetest/conformance-ipv6-ginkgo-v2.yaml
Comment (jackfrancis) [inline test/e2e/data/kubetest/conformance-ipv6-ginkgo-v2.yaml:2]: Why are we skipping `[Experimental]` in ipv6 but not dual-stack? Do `[Experimental]`-tagged test scenarios require ipv4, and our k8s dual-stack clusters are configured in such a way that those ipv4-dependent tests can be reliably tested?

File: templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml
Comment (jackfrancis) [inline templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml:159]: I assume that this `--inet4-only` flag has nothing to do w/ ipv4 k8s clusters (sort of a funny coincidence that this change would land in this PR :)).

**Review by jackfrancis — APPROVED** [review]
/lgtm

unless any of the questions I posed are actionable

File: test/e2e/data/kubetest/conformance-ipv6-ginkgo-v2.yaml
Comment (CecileRobertMichon) [reply test/e2e/data/kubetest/conformance-ipv6-ginkgo-v2.yaml:2]: Because there is no `[Experimental]` dual-stack test but there was one ipv6 one that was failing https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/4086/pull-cluster-api-provider-azure-conformance-dual-stack-with-ci-artifacts/1711771121797304320

I could add it to both but the dual-stack one would be a no-op, with ginkgo skip I prefer to only skip tests only as needed.

File: templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml
Comment (CecileRobertMichon) [reply templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml:159]: It does, this is to ensure we always connect to the ipv4 IP address of `https://storage.googleapis.com/k8s-release-dev`, IPv6 one was not reachable. It's a no-op for the ivp4 template since we always used the ipv4 adress but it ensure the ipv6/dual stack templates which are running on dual stack hosts use the ipv4 address.

Comment (jackfrancis) [reply templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml:159]: ACK, was a little strange at first to see this in the ipv6 template, but it makes sense that the actual OS is dual stack (ipv6-only config is just the k8s surface area)

**Comment by CecileRobertMichon** [comment]
@jackfrancis can you please approve if all your comments were addressed?

**Pull request merged**
